README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# autoML

<!-- badges: start -->

[![Travis build
status](https://travis-ci.com/Akai01/autoML.svg?branch=master)](https://travis-ci.com/Akai01/autoML)
[![CRAN
status](https://www.r-pkg.org/badges/version/autoML)](https://CRAN.R-project.org/package=autoML)
[![Lifecycle:
experimental](https://img.shields.io/badge/lifecycle-experimental-orange.svg)](https://www.tidyverse.org/lifecycle/#experimental)
<!-- badges: end -->

The goal of autoML is to …

## Installation

The development version of autoML can be installed from
[GitHub](https://github.com/) with:

``` r
# install.packages("devtools")
devtools::install_github("Akai01/autoML")
```

## Example

This is a basic example which shows you how to solve a common problem:

``` r
library(autoML)
## basic example code. TODO
```


NEWS.md
# autoML (development version)

# autoML 0.0.0.9000

* Added a `NEWS.md` file to track changes to the package.


tests/testthat.R
library(testthat)
library(autoML)

test_check("autoML")


tests/testthat/test-predict.R
if(require(testthat)){

  test_that("tests for some arguments in auto_xgbtree", {

    library(mlrMBO)

    data("eusilc2011", package = "autoML")

    df <- split_data(eusilc2011)

    train_data <- df$train

    test_data <- df$test

    y <- train_data$rent_ful

    train_data$rent_ful <- NULL

    x <- train_data

    x <- encode_one_hot(x)

    fit_tree <- auto_xgbtree(x, y, nrounds = list(lower = 10, upper = 12),
                             k = 3, bo_iters = 4, init_design = 20)

    fitted <- predict(fit_tree, x)

    a <- class(fitted)

    expect_that(a, equals("numeric"))

  })
}


tests/testthat/test-auto_xgbdart.R
if(require(testthat)){

  test_that("tests for some arguments in auto_xgbdart", {

    library(mlrMBO)

    data("eusilc2011", package = "autoML")

    df <- split_data(eusilc2011)

    train_data <- df$train

    test_data <- df$test

    y <- train_data$rent_ful

    train_data$rent_ful <- NULL

    x <- train_data

    x <- encode_one_hot(x)

    fit_dart <- auto_xgbdart(x, y, nrounds = list(lower = 10, upper = 12),
                             k = 3, bo_iters = 4, init_design = 25)

    a <- class(fit_dart)

    expect_that(a, equals("auto_xgb"))

  })
}


R/xgboost_cv.R
#' @importFrom xgboost xgboost xgb.DMatrix
#' @importFrom caret createFolds
#' @importFrom plyr aaply splat
xgboost_cv <- function(x, y,
                        missing = NA,
                        weight = NULL,
                        params = list(),
                        nrounds,
                        verbose = 1,
                        print_every_n = 1L,
                        early_stopping_rounds = NULL,
                        maximize = NULL,
                        save_period = NULL,
                        save_name = "xgboost.model",
                        xgb_model = NULL,
                        callbacks = list(),
                        ..., k = 10){

  folds <- caret::createFolds(1:NROW(y), k = k, list = T, returnTrain = T)
  fold_list <- data.frame("i"=1:k)

  acc <- plyr::aaply(fold_list, 1, plyr::splat(function(i) {
    train_x <- data.frame(x[folds[[i]], ])
    validation_x <- data.frame(x[-folds[[i]], ])
    train_y <- y[folds[[i]]]
    validation_y <- y[-folds[[i]]]

    if(verbose>0){
      message(paste("Training model for fold ", i))
    }

    train_x <- as.matrix(train_x)
    dtrain <- xgboost::xgb.DMatrix(train_x, label = train_y)

    fit <-    xgboost::xgboost(data = dtrain,
                               label = NULL,
                               missing = missing,
                               weight = weight,
                               params = params,
                               nrounds = nrounds,
                               verbose = verbose,
                               print_every_n = print_every_n,
                               early_stopping_rounds = early_stopping_rounds,
                               maximize = maximize,
                               save_period = save_period,
                               save_name = save_name,
                               xgb_model = xgb_model,
                               callbacks = callbacks,
                               ...)

    validation_x <- as.matrix(validation_x)

    validation_x <- xgboost::xgb.DMatrix(validation_x)

    newpred <- predict(fit, validation_x)
    acc <- accuracy(pred =  newpred, actual =  validation_y)
    acc

  }), .expand = TRUE,
  .progress = "none",
  .inform = FALSE,
  .drop = TRUE,
  .parallel = FALSE,
  .paropts = NULL
  )
  return(acc)
}


R/auto_xgbtree.R
#' Automatic model training: xgboost with tree booster.
#'
#' @description
#'
#' \Sexpr[results=rd, stage=render]{lifecycle::badge("experimental")}
#'
#' @param x A tidy data frame (samples are in rows and features are in columns)
#' containing features.
#' @param y A numeric or factor vector containing the outcome for each sample.
#' @param nrounds max number of boosting iterations. \code{0<= < Inf}
#' @param max_depth maximum depth of a tree.
#' @param eta ontrol the learning rate: scale the contribution of each tree by
#' a factor of \code{0 < eta < 1} when it is added to the current approximation.
#' Used to prevent overfitting by making the boosting process more conservative.
#' Lower value for eta implies larger value for nrounds: low eta value means
#' model more robust to overfitting but slower to compute.
#' @param gamma minimum loss reduction required to make a further partition on
#' a leaf node of the tree. The larger, the more conservative the algorithm
#' will be.
#' @param subsample subsample ratio of the training instance.
#' Setting it to 0.5 means that xgboost randomly collected half of the data
#' instances to grow trees and this will prevent overfitting.
#' It makes computation shorter (because less data to analyse).
#' It is advised to use this parameter with eta and increase nrounds.
#' @param colsample_bytree subsample ratio of columns when constructing each
#' tree.
#' @param min_child_weight minimum sum of instance weight (hessian) needed in a
#' child. If the tree partition step results in a leaf node with the sum of
#' instance weight less than min_child_weight, then the building process will
#' give up further partitioning. In linear regression mode, this simply
#' corresponds to minimum number of instances needed to be in each node.
#' The larger, the more conservative the algorithm will be.
#' @param lambda L2 regularization term on weights.
#' @param alpha L1 regularization term on weights.
#' @param missing By default is set to NA, which means that NA values should
#' be considered as 'missing' by the algorithm. Sometimes, 0 or other extreme
#' value might be used to represent missing values. This parameter is only used
#' when input is a dense matrix.
#' @param weight A vector indicating the weight for each row of the input.
#' @param verbose If 0, xgboost will stay silent. If 1, it will print
#' information about performance. If 2, some additional information will be
#' printed out. Note that setting verbose > 0 automatically engages the
#' cb.print.evaluation(period=1) callback function
#' @param print_every_n Print each n-th iteration evaluation messages when
#' verbose>0. Default is 1 which means all messages are printed.
#' This parameter is passed to the cb.print.evaluation callback.
#' @param early_stopping_rounds If NULL, the early stopping function is not
#' triggered. If set to an integer k, training with a validation set will stop
#' if the performance doesn't improve for k rounds. Setting this parameter
#' engages the cb.early.stop callback.
#' @param maximize If feval and early_stopping_rounds are set, then this
#' parameter must be set as well. When it is TRUE, it means the larger the
#' evaluation score the better. This parameter is passed to the
#' cb.early.stop callback.
#' @param save_period When it is non-NULL, model is saved to disk after every
#' save_period rounds, 0 means save at the end. The saving is handled by the
#' cb.save.model callback.
#' @param xgb_model A previously built model to continue the training from.
#' Could be either an object of class xgb.Booster, or its raw data, or the name
#'  of a file with a previously saved model.
#' @param save_name The name or path for periodically saved model file.
#' @param callbacks a list of callback functions to perform various task during
#' boosting. See callbacks. Some of the callbacks are automatically created
#' depending on the parameters' values. User can provide either existing or
#' their own callback methods in order to customize the training process.
#' @param k An integer to specify the number of folds.
#' @param validation_error_metric Later
#' @param bo_iters Number of BO iterations.
#' @param init_design Length of the initial design
#' @param \dots Other parameters to pass to \code{params}.
#' @author Resul Akay
#' @details
#'For more details please check documentation of \code{\link[xgboost]{xgboost}}.
#'
#' For more information on parameters see
#' \url{https://sites.google.com/view/lauraepp/parameters}
#'
#'
#' @return
#' An object of class \code{xgb.Booster} with the following elements:
#' \itemize{
#'   \item \code{handle} a handle (pointer) to the xgboost model in memory.
#'   \item \code{raw} a cached memory dump of the xgboost model saved as R's
#'   \code{raw} type.
#'   \item \code{niter} number of boosting iterations.
#'   \item \code{evaluation_log} evaluation history stored as a
#'   \code{data.table} with the first column corresponding to iteration number
#'   and the rest corresponding to evaluation metrics' values.
#'   It is created by the \code{\link{cb.evaluation.log}} callback.
#'   \item \code{call} a function call.
#'   \item \code{params} parameters that were passed to the xgboost library.
#'   Note that it does not capture parameters changed by the
#'   \code{\link{cb.reset.parameters}} callback.
#'   \item \code{callbacks} callback functions that were either automatically assigned or
#'   explicitly passed.
#'   \item \code{best_iteration} iteration number with the best evaluation
#'   metric value (only available with early stopping).
#'   \item \code{best_ntreelimit} the \code{ntreelimit} value corresponding to
#'   the best iteration, which could further be used in \code{predict} method
#'   (only available with early stopping).
#'   \item \code{best_score} the best evaluation metric value during early
#'   stopping. (only available with early stopping).
#'   \item \code{feature_names} names of the training dataset features
#'   (only when column names were defined in training data).
#'   \item \code{nfeatures} number of features in training data.
#' }
#'
#' @examples
#' \dontrun{
#'
#' data("eusilc2011", package = "autoML")
#'
#' df <- split_data(eusilc2011)
#'
#' train_data <- df$train
#'
#' test_data <- df$test
#'
#' y <- train_data$rent_ful
#'
#' train_data$rent_ful <- NULL
#'
#' x <- train_data
#'
#' x <- encode_one_hot(x)
#'
#' fit_dart <- auto_xgbtree(x, y, nrounds = list(lower = 10, upper = 12),
#'                          k = 3, bo_iters = 4, init_design = 20)
#' }
#' @importFrom mlrMBO makeMBOControl setMBOControlTermination mbo
#' @importFrom xgboost xgboost
#' @importFrom smoof makeSingleObjectiveFunction
#' @importFrom lhs randomLHS
#' @importFrom ParamHelpers makeParamSet makeIntegerParam makeNumericParam
#' @importFrom ParamHelpers generateDesign getParamSet
#' @export
#'
#'
auto_xgbtree <- function(x,
                      y,
                      nrounds = list(lower = 50, upper = 100),
                      max_depth = list(lower = 1, upper = 10),
                      eta = list(lower = 0.001, upper = 0.6),
                      gamma = list(lower = 0, upper = 10),
                      subsample = list(lower = 0.25, upper = 1),
                      colsample_bytree = list(lower = 0.3, upper = 0.7),
                      min_child_weight = list(lower = 1, upper = 20),
                      lambda = list(lower = 0, upper = 0.1),
                      alpha = list(lower = 0, upper = 0.1),
                      missing = NA,
                      weight = NULL,
                      verbose = 1,
                      print_every_n = 1L,
                      early_stopping_rounds = NULL,
                      maximize = NULL,
                      save_period = NULL,
                      save_name = "xgboost.model",
                      xgb_model = NULL,
                      callbacks = list(),
                      k = 5,
                      validation_error_metric = "RMSE",
                      bo_iters = 4, init_design = 15, ...){

  objective_function  <- smoof::makeSingleObjectiveFunction(
    name = "xgbtree",
    fn =   function(param){
      acc <- xgboost_cv(x, y,
                        missing = missing,
                        weight = weight,
                        params = list(
                          booster = "gbtree",
                          max_depth = param["max_depth"],
                          eta = param["eta"],
                          gamma = param["gamma"],
                          subsample = param["subsample"],
                          colsample_bytree = param["colsample_bytree"],
                          min_child_weight = param["min_child_weight"],
                          lambda = param["lambda"],
                          alpha = param["alpha"]
                                      ),
                        nrounds = param["nrounds"],
                        verbose = verbose,
                        print_every_n = print_every_n,
                        early_stopping_rounds = early_stopping_rounds,
                        maximize = maximize,
                        save_period = save_period,
                        save_name = save_name,
                        xgb_model = xgb_model,
                        callbacks = callbacks,
                        ..., k = k)

      a <- - mean(acc[,validation_error_metric])
      if(verbose>0){
        message(paste0("The mean error of ", k, " folds is ", - a))
      }
      return(a)
    },
    par.set = makeParamSet(
      makeIntegerParam("nrounds",
                       lower = nrounds[["lower"]],
                       upper = nrounds[["upper"]]),
      makeIntegerParam("max_depth",
                       lower = max_depth[["lower"]],
                       upper = max_depth[["upper"]]),
      makeNumericParam("eta",
                       lower = eta[["lower"]],
                       upper = eta[["upper"]]),
      makeNumericParam("gamma",
                       lower = gamma[["lower"]],
                       upper = gamma[["upper"]]),
      makeNumericParam("subsample",
                       lower = subsample[["lower"]],
                       upper = subsample[["upper"]]),
      makeNumericParam("colsample_bytree",
                       lower = colsample_bytree[["lower"]],
                       upper = colsample_bytree[["upper"]]),
      makeNumericParam("min_child_weight",
                       lower = min_child_weight[["lower"]],
                       upper = min_child_weight[["upper"]]),
      makeNumericParam("lambda",
                       lower = lambda[["lower"]],
                       upper = lambda[["upper"]]),
      makeNumericParam("alpha",
                       lower = alpha[["lower"]],
                       upper = alpha[["upper"]])
    ),
    minimize = FALSE
  )

  des <- ParamHelpers::generateDesign(
    n= init_design,
    par.set = ParamHelpers::getParamSet(objective_function),
    fun = lhs::randomLHS)

  control = makeMBOControl()
  control = setMBOControlTermination(control, iters = bo_iters)

  mlrmbo_result = mbo(fun = objective_function,
                      design = des,
                      control = control,
                      show.info = TRUE)
  params <- mlrmbo_result[["x"]]

  nrounds_bo <- params[["nrounds"]]

  params[["nrounds"]] <- NULL

  x <- as.matrix(x)
  dgsx <- xgboost::xgb.DMatrix(x, label = y)

  final_model <- xgboost::xgboost(data = dgsx,
                                  booster = "gbtree",
                                  label = NULL,
                                  missing = missing,
                                  weight = weight,
                                  params = params,
                                  nrounds = nrounds_bo,
                                  verbose = verbose,
                                  print_every_n = print_every_n,
                                  early_stopping_rounds = early_stopping_rounds,
                                  maximize = maximize,
                                  save_period = save_period,
                                  save_name = save_name,
                                  xgb_model = xgb_model,
                                  callbacks = callbacks,
                                  ...)
  ojct <- list("model" = final_model,"mlrmbo_result" = mlrmbo_result)

  class(ojct) <- "auto_xgb"

  return(ojct)
}


R/encode_one_hot.R
#' One-hot encoding of chategorical or caracter variables.
#'
#' @description
#'
#' \Sexpr[results=rd, stage=render]{lifecycle::badge("experimental")}
#'
#' @importFrom fastDummies dummy_columns
#' @param data A data frame.
#' @param select_columns Vector of column names that you want to create dummy
#' variables from. If NULL (default), uses all character and factor columns.
#' @param remove_first_dummy Removes the first dummy of every variable such
#' that only n-1 dummies remain. This avoids multicollinearity issues in models.
#' @param remove_most_frequent_dummy Removes the most frequently observed
#' category such that only n-1 dummies remain. If there is a tie for most
#' frequent, will remove the first (by alphabetical order) category that is
#' tied for most frequent.
#' @param ignore_na If TRUE, ignores any NA values in the column.
#' If FALSE (default), then it will make a dummy column for value_NA
#' and give a 1 in any row which has a NA value.
#' @param split A string to split a column when multiple categories are in the
#'  cell. For example, if a variable is Pets and the rows are "cat",
#'   "dog", and "turtle", each of these pets would become its own dummy column.
#'    If one row is "cat, dog", then a split value of "," this row would have a
#'    value of 1 for both the cat and dog dummy columns.
#' @param remove_selected_columns If TRUE (not default), removes the columns
#' used to generate the dummy columns.
#' @examples
#' \dontrun{
#'
#' my_data <- encode_one_hot(my_data)
#' }
#' @author Resul Akay
#' @export

encode_one_hot <- function(data,
                           remove_selected_columns = TRUE,
                           select_columns = NULL,
                           remove_first_dummy = FALSE,
                           remove_most_frequent_dummy = FALSE,
                           ignore_na = FALSE,
                           split = NULL) {

  if(!"data.frame" %in% class(data)){
    stop("data must be a tidy data.frame")
  }


data <- fastDummies::dummy_columns(data, select_columns = select_columns,
                                  remove_first_dummy = remove_first_dummy,
                                  remove_most_frequent_dummy =
                                    remove_most_frequent_dummy,
                                  ignore_na = ignore_na,
                                  split = split,
                                  remove_selected_columns =
                                    remove_selected_columns)


  return(data)
}


R/auto_xgbdart.R
#' Automatic model training: xgboost with dart booster.
#'
#' @description
#'
#' \Sexpr[results=rd, stage=render]{lifecycle::badge("experimental")}
#'
#' @param x A tidy data frame (samples are in rows and features are in columns)
#' containing features.
#' @param y A numeric or factor vector containing the outcome for each sample.
#' @param nrounds max number of boosting iterations. \code{0<= < Inf}
#' @param max_depth maximum depth of a tree.
#' @param eta ontrol the learning rate: scale the contribution of each tree by
#' a factor of \code{0 < eta < 1} when it is added to the current approximation.
#' Used to prevent overfitting by making the boosting process more conservative.
#' Lower value for eta implies larger value for nrounds: low eta value means
#' model more robust to overfitting but slower to compute.
#' @param gamma minimum loss reduction required to make a further partition on
#' a leaf node of the tree. The larger, the more conservative the algorithm
#' will be.
#' @param subsample subsample ratio of the training instance.
#' Setting it to 0.5 means that xgboost randomly collected half of the data
#' instances to grow trees and this will prevent overfitting.
#' It makes computation shorter (because less data to analyse).
#' It is advised to use this parameter with eta and increase nrounds.
#' @param colsample_bytree subsample ratio of columns when constructing each
#' tree.
#' @param min_child_weight minimum sum of instance weight (hessian) needed in a
#' child. If the tree partition step results in a leaf node with the sum of
#' instance weight less than min_child_weight, then the building process will
#' give up further partitioning. In linear regression mode, this simply
#' corresponds to minimum number of instances needed to be in each node.
#' The larger, the more conservative the algorithm will be.
#' @param lambda L2 regularization term on weights.
#' @param alpha L1 regularization term on weights.
#' @param rate_drop Dropout rate (a fraction of previous trees to drop during
#' the dropout). \code{0.0<= skip_drop <=1.0}
#' @param skip_drop Probability of skipping the dropout procedure during a
#' boosting iteration. If a dropout is skipped, new trees are added in
#' the same manner as gbtree. Note that non-zero skip_drop has higher priority
#' than rate_drop or one_drop. range: \code{0.0<= skip_drop <=1.0}
#' @param missing By default is set to NA, which means that NA values should
#' be considered as 'missing' by the algorithm. Sometimes, 0 or other extreme
#' value might be used to represent missing values. This parameter is only used
#' when input is a dense matrix.
#' @param weight A vector indicating the weight for each row of the input.
#' @param verbose If 0, xgboost will stay silent. If 1, it will print
#' information about performance. If 2, some additional information will be
#' printed out. Note that setting verbose > 0 automatically engages the
#' cb.print.evaluation(period=1) callback function
#' @param print_every_n Print each n-th iteration evaluation messages when
#' verbose>0. Default is 1 which means all messages are printed.
#' This parameter is passed to the cb.print.evaluation callback.
#' @param early_stopping_rounds If NULL, the early stopping function is not
#' triggered. If set to an integer k, training with a validation set will stop
#' if the performance doesn't improve for k rounds. Setting this parameter
#' engages the cb.early.stop callback.
#' @param maximize If feval and early_stopping_rounds are set, then this
#' parameter must be set as well. When it is TRUE, it means the larger the
#' evaluation score the better. This parameter is passed to the
#' cb.early.stop callback.
#' @param save_period When it is non-NULL, model is saved to disk after every
#' save_period rounds, 0 means save at the end. The saving is handled by the
#' cb.save.model callback.
#' @param xgb_model A previously built model to continue the training from.
#' Could be either an object of class xgb.Booster, or its raw data, or the name
#'  of a file with a previously saved model.
#' @param save_name The name or path for periodically saved model file.
#' @param callbacks a list of callback functions to perform various task during
#' boosting. See callbacks. Some of the callbacks are automatically created
#' depending on the parameters' values. User can provide either existing or
#' their own callback methods in order to customize the training process.
#' @param k An integer to specify the number of folds.
#' @param validation_error_metric Later
#' @param bo_iters Number of BO iterations.
#' @param init_design Length of the initial design
#' @param \dots Other parameters to pass to \code{params}.
#' @author Resul Akay
#' @details
#' For more details please check documentation of \code{\link[xgboost]{xgboost}}.
#'
#' For more information on parameters see
#' \url{https://sites.google.com/view/lauraepp/parameters}
#'
#' @return
#' An object of class \code{xgb.Booster} with the following elements:
#' \itemize{
#'   \item \code{handle} a handle (pointer) to the xgboost model in memory.
#'   \item \code{raw} a cached memory dump of the xgboost model saved as R's
#'   \code{raw} type.
#'   \item \code{niter} number of boosting iterations.
#'   \item \code{evaluation_log} evaluation history stored as a
#'   \code{data.table} with the first column corresponding to iteration number
#'   and the rest corresponding to evaluation metrics' values.
#'   It is created by the \code{\link{cb.evaluation.log}} callback.
#'   \item \code{call} a function call.
#'   \item \code{params} parameters that were passed to the xgboost library.
#'   Note that it does not capture parameters changed by the
#'   \code{\link{cb.reset.parameters}} callback.
#'   \item \code{callbacks} callback functions that were either automatically assigned or
#'   explicitly passed.
#'   \item \code{best_iteration} iteration number with the best evaluation
#'   metric value (only available with early stopping).
#'   \item \code{best_ntreelimit} the \code{ntreelimit} value corresponding to
#'   the best iteration, which could further be used in \code{predict} method
#'   (only available with early stopping).
#'   \item \code{best_score} the best evaluation metric value during early
#'   stopping. (only available with early stopping).
#'   \item \code{feature_names} names of the training dataset features
#'   (only when column names were defined in training data).
#'   \item \code{nfeatures} number of features in training data.
#' }
#'
#' @examples
#' \dontrun{
#'
#' data("eusilc2011", package = "autoML")
#'
#' df <- split_data(eusilc2011)
#'
#' train_data <- df$train
#'
#' test_data <- df$test
#'
#' y <- train_data$rent_ful
#'
#' train_data$rent_ful <- NULL
#'
#' x <- train_data
#'
#' x <- encode_one_hot(x)
#'
#' fit_dart <- auto_xgbdart(x, y, nrounds = list(lower = 10, upper = 12),
#'                          k = 3, bo_iters = 4, init_design = 20)
#' }
#' @importFrom mlrMBO makeMBOControl setMBOControlTermination mbo
#' @importFrom xgboost xgboost
#' @importFrom smoof makeSingleObjectiveFunction
#' @importFrom lhs randomLHS
#' @importFrom ParamHelpers makeParamSet makeIntegerParam makeNumericParam
#' @importFrom ParamHelpers generateDesign getParamSet
#' @export
#'
#'
auto_xgbdart <- function(x,
                         y,
                         nrounds = list(lower = 50, upper = 1000),
                         max_depth = list(lower = 1, upper = 10),
                         eta = list(lower = 0.001, upper = 0.6),
                         gamma = list(lower = 0, upper = 10),
                         subsample = list(lower = 0.25, upper = 1),
                         colsample_bytree = list(lower = 0.30, upper = 0.70),
                         min_child_weight = list(lower = 1, upper = 20),
                         lambda = list(lower = 0, upper = 0.1),
                         alpha = list(lower = 0, upper = 0.1),
                         rate_drop = list(lower = 0.01, upper = 0.50),
                         skip_drop = list(lower = 0.05, upper = 0.95),
                         missing = NA,
                         weight = NULL,
                         verbose = 1,
                         print_every_n = 1L,
                         early_stopping_rounds = NULL,
                         maximize = NULL,
                         save_period = NULL,
                         save_name = "xgboost.model",
                         xgb_model = NULL,
                         callbacks = list(),
                         k = 5,
                         validation_error_metric = "RMSE",
                         bo_iters = 4, init_design = 15, ...){

  objective_function  <- smoof::makeSingleObjectiveFunction(
    name = "auto_xgbdart",
    fn =   function(param){
      acc <- xgboost_cv(x, y,
                        missing = missing,
                        weight = weight,
                        params = list(
                          booster = "dart",
                          max_depth = param["max_depth"],
                          eta = param["eta"],
                          gamma = param["gamma"],
                          subsample = param["subsample"],
                          colsample_bytree = param["colsample_bytree"],
                          min_child_weight = param["min_child_weight"],
                          lambda = param["lambda"],
                          alpha = param["alpha"],
                          rate_drop = param["rate_drop"],
                          skip_drop = param["skip_drop"]
                        ),
                        nrounds = param["nrounds"],
                        verbose = verbose,
                        print_every_n = print_every_n,
                        early_stopping_rounds = early_stopping_rounds,
                        maximize = maximize,
                        save_period = save_period,
                        save_name = save_name,
                        xgb_model = xgb_model,
                        callbacks = callbacks,
                        ..., k = k)

      a <- - mean(acc[,validation_error_metric])
      if(verbose>0){
        message(paste0("The mean error of ", k, " folds is ", - a))
      }
      return(a)
    },
    par.set = ParamHelpers::makeParamSet(
      ParamHelpers::makeIntegerParam("nrounds",
                       lower = nrounds[["lower"]],
                       upper = nrounds[["upper"]]),
      ParamHelpers::makeIntegerParam("max_depth",
                       lower = max_depth[["lower"]],
                       upper = max_depth[["upper"]]),
      ParamHelpers::makeNumericParam("eta",
                       lower = eta[["lower"]],
                       upper = eta[["upper"]]),
      ParamHelpers::makeNumericParam("gamma",
                       lower = gamma[["lower"]],
                       upper = gamma[["upper"]]),
      ParamHelpers::makeNumericParam("subsample",
                       lower = subsample[["lower"]],
                       upper = subsample[["upper"]]),
      ParamHelpers::makeNumericParam("colsample_bytree",
                       lower = colsample_bytree[["lower"]],
                       upper = colsample_bytree[["upper"]]),
      ParamHelpers::makeNumericParam("min_child_weight",
                       lower = min_child_weight[["lower"]],
                       upper = min_child_weight[["upper"]]),
      ParamHelpers::makeNumericParam("lambda",
                       lower = lambda[["lower"]],
                       upper = lambda[["upper"]]),
      ParamHelpers::makeNumericParam("alpha",
                       lower = alpha[["lower"]],
                       upper = alpha[["upper"]]),
      ParamHelpers::makeNumericParam("rate_drop",
                       lower = rate_drop[["lower"]],
                       upper = rate_drop[["upper"]]),
      ParamHelpers::makeNumericParam("skip_drop",
                       lower = skip_drop[["lower"]],
                       upper = skip_drop[["upper"]])
    ),
    minimize = FALSE
  )

  des <- ParamHelpers::generateDesign(
    n= init_design,
    par.set = ParamHelpers::getParamSet(objective_function),
    fun = lhs::randomLHS)

  control = makeMBOControl()
  control = setMBOControlTermination(control, iters = bo_iters)

  mlrmbo_result = mbo(fun = objective_function,
                      design = des,
                      control = control,
                      show.info = TRUE)
  params <- mlrmbo_result[["x"]]

  nrounds_bo <- params[["nrounds"]]

  params[["nrounds"]] <- NULL

  x <- as.matrix(x)

  dgsx <- xgboost::xgb.DMatrix(x, label = y)

  final_model <- xgboost::xgboost(data = dgsx,
                                  label = NULL,
                                  missing = missing,
                                  weight = weight,
                                  booster = "dart",
                                  params = params,
                                  nrounds = nrounds_bo,
                                  verbose = verbose,
                                  print_every_n = print_every_n,
                                  early_stopping_rounds = early_stopping_rounds,
                                  maximize = maximize,
                                  save_period = save_period,
                                  save_name = save_name,
                                  xgb_model = xgb_model,
                                  callbacks = callbacks,
                                  ...)

  ojct <- list("model" = final_model,"mlrmbo_result" = mlrmbo_result)

  class(ojct) <- "auto_xgb"

  return(ojct)
}
